pkg/common/telemetry/metrics.go

```diff
@@ -31,46 +43,133 @@ type Metrics interface {
 	MeasureSinceWithLabels(key []string, start time.Time, labels []Label)
 }
 
-type MetricsConfig struct {
-	Logger      io.Writer
-	ServiceName string
-	Sinks       []Sink
-}
-
 type MetricsImpl struct {
 	*metrics.Metrics
 
-	inmemSignal *metrics.InmemSignal
+	c *MetricsConfig
+
+	promServer *http.Server
+	inmemSink  *metrics.InmemSink
 }
 
 var _ Metrics = (*MetricsImpl)(nil)
 
 // NewMetrics returns a Metric implementation
-func NewMetrics(c *MetricsConfig) *MetricsImpl {
-	// Always create an in-memory sink
-	interval := 1 * time.Second
-	retention := 1 * time.Hour
-	inmemSink := metrics.NewInmemSink(interval, retention)
+func NewMetrics(c *MetricsConfig) (*MetricsImpl, error) {
```

Comment:
IMO, this could be simplified visually, have less directly coupling to the individual implementations, and enable them to be tested separately.

```go
type sink interface {
  sink() metrics.MetricsSink
  start(ctx) error
  stop()
}
type sinkFactory func(c MetricsConfig) (sink,  error)

var sinkFactories = []sinkFactory{
  promFactory,
  dogFactory,
  statsdFactory,
}

type promSink struct {
  sink metrics.MetricsSink
  server *http.Server
  wg sync.WaitGroup
}
func promFactory(c MetricsConfig) (sink, error) {
	sink, err := prometheus.NewPrometheusSink()
	if err != nil {
		return nil, err
	}

	server := &http.Server{
		Addr:    fmt.Sprintf("localhost:%d", c.Port),
		Handler: promhttp.Handler(),
	}

       return promSink {
             sink:, sink,
            wg: sync.WaitGroup{},
            server: server,
        }, nil
}

func (s *promSink) start(ctx context.Context) error {
  s.wg.Add(1)
         go func() {
           defer wg.Done()
           promServer.ListenAndServe()
          } ()
}

func (s *promSink) stop() {
  s.server.Close()
  s.wg.Wait()
}

for _, factory := range sinkFactories {
  sink, err := factory(c)
  if err != nil {
    return nil, err
  }
  fanout = append(fanout, sink.sink())
}
```

... etc. you'll need to track the slice of `sink` interfaces as well but I think it will make this clearer. That won't compile and the spacing is whack but hopefully it makes my suggestion clear.